src/debugAdapter/goDebug.ts

```diff
@@ -321,21 +321,35 @@ class GoDebugSession extends DebugSession {
 		log('InitializeEvent');
 	}
 
+	protected findPathSeperator(path) {
+		if (/^(\w:[\\/]|\\\\)/.test(path)) return '\\';
+		return path.includes('/') ? '/' : '\\';
+	}
+
 	protected launchRequest(response: DebugProtocol.LaunchResponse, args: LaunchRequestArguments): void {
 		// Launch the Delve debugger on the program
+		let localPath = args.program;
 		let remotePath = args.remotePath || '';
 		let port = args.port || random(2000, 50000);
 		let host = args.host || '127.0.0.1';
 
 		if (remotePath.length > 0) {
-			this.localPathSeparator = args.program.includes('/') ? '/' : '\\';
-			this.remotePathSeparator = remotePath.includes('/') ? '/' : '\\';
-			if ((remotePath.endsWith('\\')) || (remotePath.endsWith('/'))) {
+			this.localPathSeparator = this.findPathSeperator(localPath);
+			this.remotePathSeparator = this.findPathSeperator(remotePath);
+
+			let llist = localPath.split(/\/|\\(?! )/).reverse();
+			let rlist = remotePath.split(/\/|\\(?! )/).reverse();
+			for(var i = 0; i < llist.length; i++) if (llist[i] !== rlist[i]) break;
+
+			if (i) {
+				localPath = llist.reverse().slice(0, -i).join(this.localPathSeparator);
+				remotePath = rlist.reverse().slice(0, -i).join(this.remotePathSeparator);
+			} else if ((remotePath.endsWith('\\')) || (remotePath.endsWith('/'))) {
 				remotePath = remotePath.substring(0, remotePath.length - 1);
 			}
 		}
 
-		this.delve = new Delve(args.mode, remotePath, port, host, args.program, args.args, args.showLog, args.cwd, args.env, args.buildFlags, args.init);
+		this.delve = new Delve(args.mode, remotePath, port, host, localPath, args.args, args.showLog, args.cwd, args.env, args.buildFlags, args.init);
 		this.delve.onstdout = (str: string) => {
 			this.sendEvent(new OutputEvent(str, 'stdout'));
 		};
```

Comment:
Remember this is initialization. The actual paths we get from remote or send to remote are the `path` arguments to `toDebuggerPath` and `toLocalPath`. Now we have four components: 
- The main module path, not visible here, but worth noting as it played a role in the old code because localPath was equal to main module path, e.g. `C:\Users\foo\go\src\foo/bar`. Note the mixed separators come from `${workspace}/bar`. Launch.json is shared between Windows & Unix.
- The `localPath`, now shortened to e.g. `C:\Users\foo`
- The `remotePath`, shortened to e.g. `/home/foo2`
- The argument to the function, e.g. `/home/foo2/go/src/foo/baz` for `toLocalPath` or (think set BP) `C:\User\foo\go\src\foo\baz` for `toDebuggerPath`.

The functions `toLocalPath` and `toDebuggerPath` just replace the beginning of the paths and change the path separators.